src/abstract_value.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+//
+use abstract_domains::AbstractDomains;
+use rpds::List;
+
+/// Mirai is an abstract interpreter and thus produces abstract values.
+/// In general, an abstract value is a value that is not fully known.
+/// For example, we may know that it is a number between 1 and 10, but not
+/// which particular number.
+///
+/// When we do know everything about a value, it is concrete rather than
+/// abstract, but is convenient to just use this structure for concrete values
+/// as well, since all operations can be uniform.
+#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Hash)]
+pub struct AbstractValue {
+    /// An abstract value is the result of some expression.
+    /// The source location of that expression is stored in provence.
+    /// When an expression is stored somewhere and then retrieved via an accessor expression, a new
+    /// abstract value is created (via refinement using the current path condition) with a provence
+    /// that is the source location of accessor expression prepended to the provence of the
+    /// refined expression.
+    pub provenance: List<Span>,
```

Comment:
Can this be empty?

And as a side note, why `List` instead of `Vec`? I'm guessing that the Rust compiler is more clever about optimizing uses of `Vec` than third-party collections, but I have no evidence of this.

Reply:
List is a functional data structure, so it enables more sharing than Vec does. Also, we will be prepending to provenance, which is not O(1) for Vec.